### ISSUE TITLE
Functional Interface inference fails inside new-style switch expression

### ISSUE THREAD

**Issue by chaosflaws** [typetools/checker-framework#5930]
This is similar in spirit to #450 and #1681.

```java
import java.util.function.Supplier;

public final class Test {
    enum TestEnum { FIRST, SECOND };

    public static void main(String[] args) {
        TestEnum testEnum = TestEnum.FIRST;
        Supplier<Integer> supplier = switch (testEnum) {
            case FIRST -> () -> 1;
            case SECOND -> () -> 2;
        };
        System.out.println(supplier.get());
    }
}
```

Parsing this java file with the checker framework (`java -jar checker.jar -processor nullness Test.java`) throws:

`error: Could not find functional interface from assignment context. Unexpected tree type: CASE For lambda tree: ()->1`

Casting the lambdas to their original type or using old-style switch statements avoid the error, but in the actual use case, both those options look rather ugly.

**Comment by smillst** [comment]
Thanks for report this!  I've fixed the crash(#5968).  I also found a false positive for switch expressions that use yield statements. I've opened a new issue for that. (#5967)
